### ISSUE TITLE
win: replace `uv_file` API

### ISSUE THREAD

**Issue by vtjnash** [libuv/libuv#856]
On Windows, a file descriptor is emulated in user-space as a wrapper around the real Win32 API. Indeed, the first operation libuv almost always must do is to throw away the wrapper (by calling `uv__get_osfhandle`). However, since the wrapper is implemented in user-space, this adds unnecessary overhead, makes the API less flexible, and is potentially error-prone (it assumes that the caller and libuv are linking against the same copy of msvcrt). A couple of functions already accept a `uv_os_sock_t`, essentially just to work around these limitations. While libuv could continuing adding dual APIs for everything (e.g. see `uv_poll_init` vs. `uv_poll_init_socket`), this doesn't seem particularly ideal to me.

What I would like to propose is to consistently remove this extra indirection from all APIs in libuv, and replace all usages of `uv_os_sock_t` with `uv_file`, while redefining `uv_file` on Windows to be a `HANDLE`.

It this is acceptable, I will start work on a PR.

Existing clients can initially transition to the new API by adding the following helper function code snippet:

``` julia
HANDLE os_handle = _get_osfhandle(fd), handle;
if (!DuplicateHandle(GetCurrentProcess(),
                     os_handle,
                     GetCurrentProcess(),
                     &handle,
                     0,
                     TRUE,
                     DUPLICATE_SAME_ACCESS)) {
  return INVALID_HANDLE_VALUE;
}
_close(fd);
return handle;
```

Longer term, those clients would now be able to transition to using the Win32 API directly instead of requiring indirection through the MSVCRT API.

**Comment by saghul** [comment]
This would definitely be something for 2.0. While your reasoning sounds ok to me, I don't have enough Windows specific knowledge to make a good judgement call here, so I'll defer to our Windows overlords @piscisaureus and @orangemocha.

**Comment by vtjnash** [comment]
Bump. Please let me know whether you would merge this.

**Comment by StefanKarpinski** [comment]
This is a fairly crucial issue for being able to deal with files on Windows. Can someone please reply to this?

**Comment by saghul** [comment]
I'll look at this more in deep by next week, sorry folks.

**Comment by saghul** [comment]
Sorry for the delay. I had a look at this and have some questions:

> However, since the wrapper is implemented in user-space, this adds unnecessary overhead, makes the API less flexible, and is potentially error-prone (it assumes that the caller and libuv are linking against the same copy of msvcrt).

I understand the multiple CRT problem, but what is the flexibility we are lacking here? Is that overhead measurable?

> What I would like to propose is to consistently remove this extra indirection from all APIs in libuv, and replace all usages of uv_os_sock_t with uv_file, while redefining uv_file on Windows to be a HANDLE.

IMHO `uv_file` is kinda a Unix-ism, since "everything is a file", I'd go for `uv_os_fd_t` or something similar. On Unix it would be an `int` and on Windows it would be a `HANDLE`. IIRC `SOCKET` can be cast to `HANDLE`, so we'd be covered for the `uv_*_open` APIs.

> Existing clients can initially transition to the new API by adding the following helper function code snippet:

I think that helper function would need to be included in libuv.

Now, how would `uv_tty_init` work? It currently gets a fd in `uv_file` form, and everyone is used to passing 0, 1, 2, standard C stdio fds. Extending this, since all fs functions would now return a `HANDLE` on Windows, how would Node translate that into an object?

Another one: when using `uv_spawn` and passing the `stdio` container, the fd of the passed handles corresponds with the index, how do we do that now? Environment variables could be used, but we are back to the previous question: how do we 'serialize' a `HANDLE`?

One possible solution is to rely on the fact that even on 64 bit systems a `HANDLE` [is actually a 32 bit integer](http://stackoverflow.com/questions/18266626/what-is-the-range-of-a-windows-handle-on-a-64-bits-application) (also [see this](http://www.viva64.com/en/k/0005/)) but I wouldn't want to base a design in some assumption that might not be future-proof.

From where I stand right now, this change would introduce tons of breakage for a not clear benefit. Can you please elaborate on the upsides of this change? Please don't take this as a discouraging note, my 'job' as a maintainer is to find all possible problems in advance, so I'd like to have a clearer picture before anyone gets to coding.

PS: If we move forward with this I'll request that you write a [LEP](https://github.com/libuv/leps) with a detailed plan and reasoning, so doing that as a response to this note could be a good thing.

**Comment by vtjnash** [comment]
> I understand the multiple CRT problem, but what is the flexibility we are lacking here? Is that overhead measurable?

For example, reference discussion at https://github.com/libuv/leps/pull/2#discussion_r21122363. In general, there are Windows API's that aren't available through the CRT. This results in the need for extra effort to, for example, write a simple implementation of `pipe`: https://github.com/libuv/libuv/blob/b12624c13693c4d29ca84b3556eadc9e9c0936a4/test/test-spawn.c#L1568

> IIRC SOCKET can be cast to HANDLE

Yes, they're the same size. (although the WSA API is completely independent, so they mostly aren't interchangeable)

> Now, how would uv_tty_init work? It currently gets a fd in uv_file form, and everyone is used to passing 0, 1, 2, standard C stdio fds. Extending this, since all fs functions would now return a HANDLE on Windows, how would Node translate that into an object?

The STDIO handles on windows are -10, -11, and -12. Unlike the standard CRT handles (0,1,2), it is not undefined behavior to access these from a graphical program (https://support.microsoft.com/en-us/kb/105305). The application can also use GetStdHandle https://msdn.microsoft.com/en-us/library/windows/desktop/ms683231(v=vs.85).aspx to retrieve the real handle instead of the pseudo handle if it wants a stable / non-redirectable handle.

> Another one: when using uv_spawn and passing the stdio container, the fd of the passed handles corresponds with the index, how do we do that now? Environment variables could be used, but we are back to the previous question: how do we 'serialize' a HANDLE?

We do that now by extracting the HANDLE from the fd and passing the first 3 as the StdInput, StdOutput, and StdError Handles, respectively. (The Win32 API only supports passing 3 HANDLEs, and they are named, not indexed).

> PS: If we move forward with this I'll request that you write a LEP with a detailed plan and reasoning, so doing that as a response to this note could be a good thing.

Sounds good. I'll start working on that too. I don't mind discussion, it's silence that's annoying.

**Comment by saghul** [comment]
> The STDIO handles on windows are -10, -11, and -12. Unlike the standard CRT handles (0,1,2), it is not undefined behavior to access these from a graphical program (https://support.microsoft.com/en-us/kb/105305). The application can also use GetStdHandle https://msdn.microsoft.com/en-us/library/windows/desktop/ms683231(v=vs.85).aspx to retrieve the real handle instead of the pseudo handle if it wants a stable / non-redirectable handle.

That's good. We could have some defines a la `UV_STDIN_FD` et al, which map to `STDIN_FILENO` on unix and those you mentioned on Windows.

> We do that now by extracting the HANDLE from the fd and passing the first 3 as the StdInput, StdOutput, and StdError Handles, respectively. (The Win32 API only supports passing 3 HANDLEs, and they are named, not indexed).

I meant in the child. Or maybe I'm confused and it's a Unix thing? AFAIS Node use NODE_CHANNEL_FD environment variable for the children get the fd and can communicate with the parent: https://github.com/nodejs/node/blob/master/lib/internal/child_process.js#L269-L273 the fd number seems to match the index in the stdio array.

> Sounds good. I'll start working on that too. I don't mind discussion, it's silence that's annoying.

👍

**Comment by vtjnash** [comment]
> I meant in the child. Or maybe I'm confused and it's a Unix thing? AFAIS Node use NODE_CHANNEL_FD environment variable for the children get the fd and can communicate with the parent: https://github.com/nodejs/node/blob/master/lib/internal/child_process.js#L269-L273 the fd number seems to match the index in the stdio array.

This is modeled after the Unix thing. It's possible to pass more than 3 handles to a child via a call to DuplicateHandle (https://msdn.microsoft.com/en-us/library/windows/desktop/ms724251(v=vs.85).aspx) after the child has started, but any `fd` number would be assigned by the mscvrt library in the child only if you called `_open_osfhandle` on the handle. There's no way to map a fd across to an arbitrary number. However, the fds 0, 1, and 2 are special, since they are automatically mapped to the pseudo-handles -10, -11, and -12 during crt initialization. Therefore, at launch, these are roughly interchangeable ids.

**Comment by saghul** [comment]
> This is modeled after the Unix thing. It's possible to pass more than 3 handles to a child via a call to DuplicateHandle (https://msdn.microsoft.com/en-us/library/windows/desktop/ms724251(v=vs.85).aspx) after the child has started, but any fd number would be assigned by the mscvrt library in the child only if you called _open_osfhandle on the handle. There's no way to map a fd across to an arbitrary number. However, the fds 0, 1, and 2 are special, since they are automatically mapped to the pseudo-handles -10, -11, and -12 during crt initialization. Therefore, at launch, these are roughly interchangeable ids.

I see. My question is about how we would do it after the change. Now, AFAIS, because we use CRT fds they are allocated in order, and we know the 4th entry in the stdio array will have fd 3, so one can export a variable indicating this fact, but if we switch to using HANDLE, this won't be predictable, and it's not yet clear to me how we would signal the child that there is an IPC channel setup.

**Comment by vtjnash** [comment]
I discovered I misspoke: there is no \*documented\* way to pass a fd to the child, so I didn't realize libuv is using a little trick. For a child program compiled against the Microsoft C runtime, there's a magic buffer (Reserved2) which can be set to contain a list of the HANDLEs. These HANDLEs will be reopened as file descriptors by the child CRT in the order they are given in the list.

**Comment by d3x0r** [comment]
I have a different reason to want it to use more of a bare windows API instead of temporary storage in C runtime.  Node is linked with a static runtime, so no mixed runtime DLL issues; but that means that the call that is   ```_open_osfhandle``` just makes a windows HANDLE into an int for internal tracking and then ```handle = uv__get_osfhandle(fd);``` is done in most places to get back to the HANDLE.  This ends up doing a validation in the first place (open_osfhandle) becomes static code that REQUIRES the HANDLE be a real HANDLE instead of just trusting that it is.  So right now, for me, even Open File is crashing; because 

there seems to not really be a C API use of the fd, except to get the handle back out of it.

And the above DuplicateHandle would be redundant.  It's not the handle duplication that's required... it's just converting from HANDLE to 'int' which might be smaller, and fits more with the posix way of things; but this tracking and assignment of integer to handle COULD be a  libUV lookup instead.  Then at least if I give back a handle from CreateFileW that is X it might not try to assume it's really really REALLY a HANDLE.

**Comment by vtjnash** [comment]
Implemented in #1166.
